### ISSUE TITLE
[Feature Request] Save trained classifier

### ISSUE THREAD

**Issue by geblanco** [autogoal/autogoal#43]
Hi!

After fitting to find the best pipeline, you can save/load it with the appropriate methods, but this only saves the best pipeline found yet.

It would be a nice feature to be able to save the fitted classifier, otherwise, the best pipeline has to be retrained every time, which is very time consuming.

I have tried to do something similar with:
```python
best_pipe = classifier.best_pipeline_.sampler_.replay()
pickle.dump(best_pipe, open('./trained_classifier.pkl', 'wb'))
```

But it fails to load it again later with a maximum recursion limit error. Raising the recursion limit doesn't solve it either.

What do you think? Would this be a desirable feature or is out of scope?

Best,

**Comment by apiad** [comment]
Yeah, this is a major request since day one. The reason why this fails is that we generate a bunch of dynamic classes (all tuple and list wrappers) and they are not trivial to serialize.

The good news is that pull request #20 solves this already, with unit tests and all!!! So once that is merged, this will get closed as well.

Thanks for asking :+1: !!

**Comment by geblanco** [comment]
Excellent @apiad, I'll try it right away! :)

**Comment by geblanco** [comment]
Hello!

I have finally tried out this functionality and found a problem.
Saving/restoring the object seems correct (equal representation, same parameter values?), but when using the restored pipeline to predict, I get a `NotFittedError` (the models' weights are not saved, I guess)

For example, using plain sklearn's LogisticRegression, you instantiate the model, fit it and then predict, afterwards, doing a pickle.dump/load is safe (avoiding the infamous `NotFittedError`), effectively saving the model's parameters and restoring them later.

Following the same procedure inside AutoGOAL yields a pipeline unable to predict, throwing the `NotFittedError` on every model. By now, I don't have a MWE, but maybe I can craft one soon.

What do you think?

**Comment by apiad** [comment]
Go for it! :+1: If you want to work on this, I would suggest opening a new PR for that specific discussion.

**Comment by geblanco** [comment]
Hello @apiad, excellent! I'm a bit lost with dynamic generation of classes in python (and this lib), maybe you can enlighten me a bit as to where to start from and I can submit a PR!

**Comment by geblanco** [comment]
Hi,

I've finished the MWE, and it seems to work perfectly, I guess the bug is in my code, I'm looking into it right now and think it's on the save part.

In any case, here is the MWE, just in case it's useful for the future somehow:

```python
####### IMPORTS #######
# disable warnings
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

import numpy as np

from pickle import Pickler, Unpickler
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_digits
from sklearn.pipeline import Pipeline as SkPipeline
from sklearn.linear_model import LogisticRegression as LogisticRegression
from sklearn.model_selection import train_test_split

from autogoal.ml import AutoML
from autogoal.search import PESearch
from autogoal.contrib.sklearn import LogisticRegression as SkLogisticRegression
from autogoal.kb import (
    MatrixContinuousDense,
    CategoricalVector,
)
from autogoal.grammar import (
    Union,
    generate_cfg,
    Categorical,
    Continuous,
)

####### DATA VARS #######
sklearn_dump_file = "./test_lr.pkl"
pe_search_dump_file = "./test_pe.pkl"
automl_dump_file = "./test_automl.pkl"
iterations = 10
pop_size = 5
pop_selection = 1
early_stop = 5

digits = load_digits()
x_train, x_test, y_train, y_test = train_test_split(
    digits.data, digits.target, test_size=0.25, random_state=0
)

####### UTIL & FUNCTIONS #######
def check_restore(file_path, x_test, prev_preds, cls):
    restored = Unpickler(open(file_path, "rb")).load()
    assert(isinstance(restored, cls))
    predictions_restored = restored.predict(x_test)
    assert(np.all(prev_preds == predictions_restored))

class LR(SkLogisticRegression):
    def __init__(
        self,
        penalty: Categorical("l1", "l2"),
        reg: Continuous(0.1, 10),
    ):
        super(LR, self).__init__(
            penalty=penalty,
            C=reg,
            solver="liblinear",
            fit_intercept=False,
            multi_class="auto",
            dual=False,
        )
        self.penalty = penalty
        self.reg = reg

class LogRegPipeline(SkPipeline):
    def __init__(
        self,
        classifier: Union("Classifier", LR),
    ):
        self.classifier = classifier
        super(LogRegPipeline, self).__init__([
            ("class", classifier),
        ])

def make_fn(X_train, y_train, X_test, y_test, score_fn):
    def fitness(pipeline):
        pipeline.fit(X_train, y_train)
        y_pred = pipeline.predict(X_test)
        return score_fn(y_test, y_pred)

    return fitness

####### SKLEARN TEST #######
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
sk_preds = logreg.predict(x_test)

# save the model
Pickler(open(sklearn_dump_file, "wb")).dump(logreg)
# restore
check_restore(sklearn_dump_file, x_test, sk_preds, LogisticRegression)

####### AUTOGOAL TEST #######
grammar = generate_cfg(LogRegPipeline)
fitness_fn = make_fn(x_train, y_train, x_test, y_test, accuracy_score)

####### PE Algorithm #######
pe_algo = PESearch(
    grammar,
    fitness_fn,
    pop_size=pop_size,
    selection=pop_selection,
    early_stop=early_stop,
)
pe_best, score = pe_algo.run(iterations)
pe_best.fit(x_train, y_train)
pe_preds = pe_best.predict(x_test)
Pickler(open(pe_search_dump_file, "wb")).dump(pe_best)
check_restore(pe_search_dump_file, x_test, pe_preds, LogRegPipeline)

####### AUTOML TEST #######
automl = AutoML(
    input=MatrixContinuousDense(),
    output=CategoricalVector(),
    search_algorithm=PESearch,
    search_iterations=iterations,
    score_metric=accuracy_score,
    search_kwargs=dict(
        pop_size=pop_size,
        selection=pop_selection,
        early_stop=early_stop,
    ),
)
score = fitness_fn(automl)
automl_preds = automl.predict(x_test)
automl.save(open(automl_dump_file, "wb"))
check_restore(automl_dump_file, x_test, automl_preds, AutoML)
```

Thanks,
Best regards
